pstoreds/addr_book.go

```diff
@@ -2,398 +2,338 @@ package pstoreds
+	// /peers/addrs/<b32 peer id no padding>
+	addrBookBase = ds.NewKey("/peers/addrs")
+)
 
-func (ar *addrRecord) UnmarshalBinary(b []byte) error {
-	ar.ttl = time.Duration(binary.LittleEndian.Uint64(b))
-	// this had been serialized by us, no need to check for errors
-	ar.addr, _ = ma.NewMultiaddrBytes(b[8:])
-	return nil
+// addrsRecord decorates the AddrBookRecord with locks and metadata.
+type addrsRecord struct {
+	sync.RWMutex
+	*pb.AddrBookRecord
+	dirty bool
 }
 
-// NewAddrBook initializes a new address book given a
-// Datastore instance, a context for managing the TTL manager,
-// and the interval at which the TTL manager should sweep the Datastore.
-func NewAddrBook(ctx context.Context, store ds.TxnDatastore, opts Options) (*dsAddrBook, error) {
-	if _, ok := store.(ds.TTLDatastore); !ok {
-		return nil, ErrTTLDatastore
+// flush writes the record to the datastore by calling ds.Put, unless the record is
+// marked for deletion, in which case we call ds.Delete. To be called within a lock.
+func (r *addrsRecord) flush(write ds.Write) (err error) {
+	key := addrBookBase.ChildString(b32.RawStdEncoding.EncodeToString([]byte(r.Id.ID)))
+	if len(r.Addrs) == 0 {
+		return write.Delete(key)
 	}
 
-	var (
-		cache cache = &noopCache{}
-		err   error
-	)
-
-	if opts.CacheSize > 0 {
-		if cache, err = lru.NewARC(int(opts.CacheSize)); err != nil {
-			return nil, err
-		}
+	data, err := r.Marshal()
+	if err != nil {
+		return err
 	}
-
-	mgr := &dsAddrBook{
-		cache:        cache,
-		ds:           store,
-		subsManager:  pstoremem.NewAddrSubManager(),
-		writeRetries: int(opts.WriteRetries),
+	if err = write.Put(key, data); err != nil {
+		return err
 	}
-	return mgr, nil
+	// write succeeded; record is no longer dirty.
+	r.dirty = false
+	return nil
 }
 
-func keysAndAddrs(p peer.ID, addrs []ma.Multiaddr) ([]ds.Key, []ma.Multiaddr, error) {
-	var (
-		keys      = make([]ds.Key, len(addrs))
-		clean     = make([]ma.Multiaddr, len(addrs))
-		parentKey = abBase.ChildString(base32.RawStdEncoding.EncodeToString([]byte(p)))
-		i         = 0
-	)
-
-	for _, addr := range addrs {
-		if addr == nil {
-			continue
-		}
-
-		hash, err := mh.Sum((addr).Bytes(), mh.MURMUR3, -1)
-		if err != nil {
-			return nil, nil, err
+// clean is called on records to perform housekeeping. The return value indicates if the record was changed
+// as a result of this call.
+//
+// clean does the following:
+// * sorts addresses by expiration (soonest expiring first).
+// * removes expired addresses.
+//
+// It short-circuits optimistically when there's nothing to do.
+//
+// clean is called from several points:
+// * when accessing an entry.
+// * when performing periodic GC.
+// * after an entry has been modified (e.g. addresses have been added or removed, TTLs updated, etc.)
+//
+// If the return value is true, the caller can perform a flush immediately, or can schedule an async
+// flush, depending on the context.
+func (r *addrsRecord) clean() (chgd bool) {
```

Comment:
It seems to me that `clean` should be calling `r.Lock()` or all users for `clean` should be doing it.

Reply:
So `UpdateAddrs` calls `clean()` and locks it beforehand. We should lock it also after fetching it from the cache and calling `clean()`.

Reply:
Lack of reentrant locks in Go is a PITA.

Reply:
> We should lock it also after fetching it from the cache and calling `clean()`.

Done ✅